system/lib/wasmfs/file.h

```diff
@@ -79,10 +82,19 @@ class File : public std::enable_shared_from_this<File> {
 
   Handle locked() { return Handle(shared_from_this()); }
 
+  std::optional<Handle> maybeLocked() {
+    auto handle = Handle(shared_from_this(), std::defer_lock);
+    if (handle.trylock()) {
+      return Handle(shared_from_this());
+    } else {
+      return {};
+    }
+  }
+
 protected:
   File(FileKind kind, mode_t mode) : kind(kind), mode(mode) {}
   // A mutex is needed for multiple accesses to the same file.
-  std::mutex mutex;
+  std::recursive_mutex mutex;
```

Comment:
When we have benchmarks it would be good to see what effect this has (we will be able to flip this to a normal mutex so long as the benchmark doesn't rename).

Reply:
A non-trivial comment explaining why this mutex is recursive is necessary here, I think.